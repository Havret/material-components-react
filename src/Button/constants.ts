const BUTTON = 'mdc-button';
const BUTTON_ICON = 'mdc-button__icon';
const BUTTON_RAISED = 'mdc-button--raised';
const BUTTON_UNELEVATED = 'mdc-button--unelevated';
const BUTTON_STROKED = 'mdc-button--stroked';
const BUTTON_DENSE = 'mdc-button--dense';
const BUTTON_COMPACT = 'mdc-button--compact';

export { BUTTON, BUTTON_ICON, BUTTON_RAISED, BUTTON_UNELEVATED, BUTTON_STROKED, BUTTON_DENSE, BUTTON_COMPACT }