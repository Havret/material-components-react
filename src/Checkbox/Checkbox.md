Checkboxes allow the user to select multiple options from a set.

## Ripple enabled

```js
initialState = { checked: false };
<Checkbox ripple checked={state.checked} onChange={e => setState({checked: e.target.checked})} />
```

## CSS Only

### Default checkbox

```js
initialState = { checked: false };
<Checkbox checked={state.checked} onChange={e => setState({checked: e.target.checked})} />
```

### Disabled checkbox

```js
initialState = { checked: false, disabled: true };
<div>
<Checkbox checked={state.checked} disabled={state.disabled} onChange={e => setState({checked: e.target.checked})} />
<Button stroked onClick={() => setState(prev => ({disabled: !prev.disabled}))}>Toggle Disabled</Button>
</div>
```

### Indeterminate checkbox

```js
initialState = { checked: false, indeterminate: true };

<div>
<Checkbox checked={state.checked} indeterminate={state.indeterminate} onChange={e => setState({checked: e.target.checked, indeterminate: false})} />
<Button stroked onClick={() => setState(prev => ({indeterminate: !prev.indeterminate}))}>Toggle Indeterminate</Button>
</div>
```