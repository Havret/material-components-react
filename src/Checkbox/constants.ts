export const ROOT = 'mdc-checkbox';
export const UPGRADED = 'mdc-checkbox--upgraded';
export const NATIVE_CONTROL = 'mdc-checkbox__native-control';
export const DISABLED = 'mdc-checkbox--disabled';
export const BACKGROUND = 'mdc-checkbox__background';
export const CHECKMARK = 'mdc-checkbox__checkmark';
export const CHECKMARK_PATH = 'mdc-checkbox__checkmark__path';
export const MIXEDMARK = 'mdc-checkbox__mixedmark';

export const ANIM_UNCHECKED_CHECKED = 'mdc-checkbox--anim-unchecked-checked';
export const ANIM_UNCHECKED_INDETERMINATE = 'mdc-checkbox--anim-unchecked-indeterminate';
export const ANIM_CHECKED_UNCHECKED = 'mdc-checkbox--anim-checked-unchecked';
export const ANIM_CHECKED_INDETERMINATE = 'mdc-checkbox--anim-checked-indeterminate';
export const ANIM_INDETERMINATE_CHECKED = 'mdc-checkbox--anim-indeterminate-checked';
export const ANIM_INDETERMINATE_UNCHECKED = 'mdc-checkbox--anim-indeterminate-unchecked';

export const ANIM_END_LATCH_MS = 100;

export enum CheckState {
    checked,
    unchecked,
    indeterminate
}