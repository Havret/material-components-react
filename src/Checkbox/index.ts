export { default, Checkbox, CheckboxProps } from './Checkbox';