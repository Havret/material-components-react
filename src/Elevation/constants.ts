export const ELEVATION = 'mdc-elevation--z';
export const ELEVATION_TRANSITION = 'mdc-elevation-transition';