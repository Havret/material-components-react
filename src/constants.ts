export const MATERIAL_ICONS = 'material-icons';