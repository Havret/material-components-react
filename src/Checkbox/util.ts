import {
    ANIM_CHECKED_INDETERMINATE,
    ANIM_CHECKED_UNCHECKED,
    ANIM_INDETERMINATE_CHECKED,
    ANIM_INDETERMINATE_UNCHECKED,
    ANIM_UNCHECKED_CHECKED,
    ANIM_UNCHECKED_INDETERMINATE,
    CheckState
} from './constants';
import { CheckboxProps } from './Checkbox';

function determineCheckState(nextProps: Partial<CheckboxProps>): CheckState {
    if (nextProps.indeterminate) {
        return CheckState.indeterminate;
    }

    return nextProps.checked ? CheckState.checked : CheckState.unchecked;
}

function getTransitionAnimationClass(oldCheckState: CheckState, newCheckState: CheckState): string {
    switch (oldCheckState) {
        case CheckState.unchecked:
            return newCheckState === CheckState.checked ? ANIM_UNCHECKED_CHECKED : ANIM_UNCHECKED_INDETERMINATE;
        case CheckState.checked:
            return newCheckState === CheckState.unchecked ? ANIM_CHECKED_UNCHECKED : ANIM_CHECKED_INDETERMINATE;
        // CheckState.indeterminate:
        default:
            return newCheckState === CheckState.checked ? ANIM_INDETERMINATE_CHECKED : ANIM_INDETERMINATE_UNCHECKED;
    }
}

export { determineCheckState, getTransitionAnimationClass };