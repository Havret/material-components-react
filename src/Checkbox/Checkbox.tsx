import * as cx from 'classnames';
import * as React from 'react';
import { HTMLProps, ChangeEventHandler } from 'react';

import { ObjectOmit } from '../utils';
import {
    ANIM_END_LATCH_MS,
    BACKGROUND,
    CHECKMARK,
    CHECKMARK_PATH,
    DISABLED,
    MIXEDMARK,
    NATIVE_CONTROL,
    ROOT,
    UPGRADED,
    CheckState
} from './constants';
import { determineCheckState, getTransitionAnimationClass } from './util';

interface CheckboxProps extends ObjectOmit<HTMLProps<HTMLInputElement>, 'type'> {
    /**
     * Addition CSS classes that will be added to the root element.
     */
    className?: string;
    /**
     * If `true`, the checkbox is checked.
     */
    checked?: boolean;
    /**
     * If `true`, the checkbox is indeterminate.
     */
    indeterminate?: boolean;
    /**
     * If `true`, the checkbox will be disabled.
     */
    disabled?: boolean;
    /**
     *    Callback fired when the state is changed.
     */
    onChange?: ChangeEventHandler<HTMLInputElement>;
}

interface CheckboxState {
    checkState: CheckState;
    transitionAnimationClass: string;
}

class Checkbox extends React.Component<CheckboxProps, CheckboxState> {
    static defaultProps = {
        checked: false,
        indeterminate: false,
        disabled: false,
    };

    // https://github.com/Microsoft/TypeScript/issues/842
    // tslint:disable-next-line:no-any
    private _animEndLatchTimer: any = 0;

    constructor(props: CheckboxProps) {
        super(props);

        this.state = {
            checkState: determineCheckState(props),
            transitionAnimationClass: ''
        };
    }

    componentWillReceiveProps(nextProps: Readonly<CheckboxProps>): void {
        const checkState = determineCheckState(nextProps);
        const oldCheckState = this.state.checkState;

        if (checkState === this.state.checkState) {
            return;
        }

        const transitionAnimationClass = getTransitionAnimationClass(oldCheckState, checkState);

        this.setState({checkState, transitionAnimationClass});
    }

    render() {
        const {
            className,
            disabled,
            indeterminate,  // Filter out properties which should not
            checked,        // be passed to native input element
            ...rest
        } = this.props;
        const {transitionAnimationClass, checkState} = this.state;

        const classNames = cx(ROOT, UPGRADED, transitionAnimationClass, {
            [DISABLED]: disabled
        }, className);

        const onAnimationEnd = transitionAnimationClass.length > 0
            ? this.handleOnAnimationEnd
            : undefined;

        return (
            <div
                className={classNames}
                onAnimationEnd={onAnimationEnd}
            >
                <input
                    {...rest}
                    type="checkbox"
                    className={NATIVE_CONTROL}
                    disabled={disabled}
                    checked={checkState === CheckState.checked}
                    ref={x => x ? x.indeterminate = checkState === CheckState.indeterminate : undefined}
                />
                <div className={BACKGROUND}>
                    <svg className={CHECKMARK} viewBox="0 0 24 24">
                        <path
                            className={CHECKMARK_PATH}
                            fill="none"
                            stroke="white"
                            d="M1.73,12.91 8.1,19.28 22.79,4.59"
                        />
                    </svg>
                    <div className={MIXEDMARK}/>
                </div>
            </div>
        );
    }

    private handleOnAnimationEnd: React.AnimationEventHandler<HTMLDivElement> = () => {
        clearTimeout(this._animEndLatchTimer);
        this._animEndLatchTimer = setTimeout(() => {
            this.setState({transitionAnimationClass: ''});
        }, ANIM_END_LATCH_MS);
    };
}

export { Checkbox as default, Checkbox, CheckboxProps };