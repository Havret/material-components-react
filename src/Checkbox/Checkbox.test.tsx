import { configure, shallow, mount } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import * as React from 'react';
import toJson from 'enzyme-to-json';

import { Checkbox } from './Checkbox';
import {
    ANIM_CHECKED_INDETERMINATE,
    ANIM_CHECKED_UNCHECKED,
    ANIM_INDETERMINATE_CHECKED,
    ANIM_INDETERMINATE_UNCHECKED,
    ANIM_UNCHECKED_CHECKED,
    ANIM_UNCHECKED_INDETERMINATE,
    DISABLED, ROOT
} from './constants';
import { cssClasses as rippleCssClasses } from '../Ripple/constants';

describe('Checkbox', () => {
    configure({adapter: new Adapter()});

    test('It should have mdc classes - Ripple Disabled', () => {
        const wrapper = shallow(<Checkbox ripple={false}/>);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('It should have mdc classes - Ripple Enabled', () => {
        const wrapper = mount(<Checkbox/>);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('It should have disabled class', () => {
        const wrapper = shallow(<Checkbox disabled={true} ripple={false}/>);
        expect(wrapper.hasClass(DISABLED)).toBeTruthy();
    });

    test('It should add extra class to the root element', () => {
        const extraClass = 'extra-class';
        const wrapper = shallow(<Checkbox className={extraClass} ripple={false}/>);
        expect(wrapper.hasClass(extraClass)).toBeTruthy();
    });

    test('It should be initialized without transition animation classes', () => {
        const wrapper = shallow(<Checkbox ripple={false}/>);

        expect(wrapper.hasClass(ANIM_CHECKED_INDETERMINATE)).toBeFalsy();
        expect(wrapper.hasClass(ANIM_CHECKED_UNCHECKED)).toBeFalsy();
        expect(wrapper.hasClass(ANIM_INDETERMINATE_CHECKED)).toBeFalsy();
        expect(wrapper.hasClass(ANIM_INDETERMINATE_UNCHECKED)).toBeFalsy();
        expect(wrapper.hasClass(ANIM_UNCHECKED_CHECKED)).toBeFalsy();
        expect(wrapper.hasClass(ANIM_UNCHECKED_INDETERMINATE)).toBeFalsy();
    });

    test('It should set transition animation class after state changed', () => {
        const wrapper = mount(<Checkbox/>);
        wrapper.setProps({checked: true});
        expect(wrapper.find(`.${ROOT}`).hasClass(ANIM_UNCHECKED_CHECKED)).toBeTruthy();
    });

    test('It should reset transition animation class after transition ends', () => {
        // arrange
        jest.useFakeTimers();
        const wrapper = shallow(<Checkbox/>);
        wrapper.setProps({checked: true});

        // act
        wrapper.simulate('animationEnd');
        jest.runAllTimers();
        wrapper.update();

        // assert
        expect(wrapper.hasClass(ANIM_UNCHECKED_CHECKED)).toBeFalsy();
    });

    test('It should be ripple upgraded when when ripple enabled', () => {
        const wrapper = mount(<Checkbox ripple={true}/>);
        expect(wrapper.find(`.${ROOT}`).hasClass(rippleCssClasses.ROOT)).toBeTruthy();
    });
});