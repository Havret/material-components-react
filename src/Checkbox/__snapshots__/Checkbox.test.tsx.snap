// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox It should have mdc classes 1`] = `
<div
  className="mdc-checkbox mdc-checkbox--upgraded"
>
  <input
    checked={false}
    className="mdc-checkbox__native-control"
    disabled={false}
    ripple={false}
    type="checkbox"
  />
  <div
    className="mdc-checkbox__background"
  >
    <svg
      className="mdc-checkbox__checkmark"
      viewBox="0 0 24 24"
    >
      <path
        className="mdc-checkbox__checkmark__path"
        d="M1.73,12.91 8.1,19.28 22.79,4.59"
        fill="none"
        stroke="white"
      />
    </svg>
    <div
      className="mdc-checkbox__mixedmark"
    />
  </div>
</div>
`;
